// nestjs-proto-gen-ts/dist/vendor/google/protobuf/compiler/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.google = void 0;
var google;
(function (google) {
    let protobuf;
    (function (protobuf) {
        let FieldDescriptorProto;
        (function (FieldDescriptorProto) {
            let Type;
            (function (Type) {
                Type[Type["TYPE_DOUBLE"] = 1] = "TYPE_DOUBLE";
                Type[Type["TYPE_FLOAT"] = 2] = "TYPE_FLOAT";
                Type[Type["TYPE_INT64"] = 3] = "TYPE_INT64";
                Type[Type["TYPE_UINT64"] = 4] = "TYPE_UINT64";
                Type[Type["TYPE_INT32"] = 5] = "TYPE_INT32";
                Type[Type["TYPE_FIXED64"] = 6] = "TYPE_FIXED64";
                Type[Type["TYPE_FIXED32"] = 7] = "TYPE_FIXED32";
                Type[Type["TYPE_BOOL"] = 8] = "TYPE_BOOL";
                Type[Type["TYPE_STRING"] = 9] = "TYPE_STRING";
                Type[Type["TYPE_GROUP"] = 10] = "TYPE_GROUP";
                Type[Type["TYPE_MESSAGE"] = 11] = "TYPE_MESSAGE";
                Type[Type["TYPE_BYTES"] = 12] = "TYPE_BYTES";
                Type[Type["TYPE_UINT32"] = 13] = "TYPE_UINT32";
                Type[Type["TYPE_ENUM"] = 14] = "TYPE_ENUM";
                Type[Type["TYPE_SFIXED32"] = 15] = "TYPE_SFIXED32";
                Type[Type["TYPE_SFIXED64"] = 16] = "TYPE_SFIXED64";
                Type[Type["TYPE_SINT32"] = 17] = "TYPE_SINT32";
                Type[Type["TYPE_SINT64"] = 18] = "TYPE_SINT64";
            })(Type = FieldDescriptorProto.Type || (FieldDescriptorProto.Type = {}));
            let Label;
            (function (Label) {
                Label[Label["LABEL_OPTIONAL"] = 1] = "LABEL_OPTIONAL";
                Label[Label["LABEL_REQUIRED"] = 2] = "LABEL_REQUIRED";
                Label[Label["LABEL_REPEATED"] = 3] = "LABEL_REPEATED";
            })(Label = FieldDescriptorProto.Label || (FieldDescriptorProto.Label = {}));
        })(FieldDescriptorProto = protobuf.FieldDescriptorProto || (protobuf.FieldDescriptorProto = {}));
        let FileOptions;
        (function (FileOptions) {
            let OptimizeMode;
            (function (OptimizeMode) {
                OptimizeMode[OptimizeMode["SPEED"] = 1] = "SPEED";
                OptimizeMode[OptimizeMode["CODE_SIZE"] = 2] = "CODE_SIZE";
                OptimizeMode[OptimizeMode["LITE_RUNTIME"] = 3] = "LITE_RUNTIME";
            })(OptimizeMode = FileOptions.OptimizeMode || (FileOptions.OptimizeMode = {}));
        })(FileOptions = protobuf.FileOptions || (protobuf.FileOptions = {}));
        let FieldOptions;
        (function (FieldOptions) {
            let CType;
            (function (CType) {
                CType[CType["STRING"] = 0] = "STRING";
                CType[CType["CORD"] = 1] = "CORD";
                CType[CType["STRING_PIECE"] = 2] = "STRING_PIECE";
            })(CType = FieldOptions.CType || (FieldOptions.CType = {}));
            let JSType;
            (function (JSType) {
                JSType[JSType["JS_NORMAL"] = 0] = "JS_NORMAL";
                JSType[JSType["JS_STRING"] = 1] = "JS_STRING";
                JSType[JSType["JS_NUMBER"] = 2] = "JS_NUMBER";
            })(JSType = FieldOptions.JSType || (FieldOptions.JSType = {}));
        })(FieldOptions = protobuf.FieldOptions || (protobuf.FieldOptions = {}));
        let MethodOptions;
        (function (MethodOptions) {
            let IdempotencyLevel;
            (function (IdempotencyLevel) {
                IdempotencyLevel[IdempotencyLevel["IDEMPOTENCY_UNKNOWN"] = 0] = "IDEMPOTENCY_UNKNOWN";
                IdempotencyLevel[IdempotencyLevel["NO_SIDE_EFFECTS"] = 1] = "NO_SIDE_EFFECTS";
                IdempotencyLevel[IdempotencyLevel["IDEMPOTENT"] = 2] = "IDEMPOTENT";
            })(IdempotencyLevel = MethodOptions.IdempotencyLevel || (MethodOptions.IdempotencyLevel = {}));
        })(MethodOptions = protobuf.MethodOptions || (protobuf.MethodOptions = {}));
    })(protobuf = google.protobuf || (google.protobuf = {}));
})(google = exports.google || (exports.google = {}));
