// ingreedie/src/parser.js
module.exports = (function() {
  "use strict";

  /*
   * Generated by PEG.js 0.9.0.
   *
   * http://pegjs.org/
   */

  function peg$subclass(child, parent) {
    function ctor() { this.constructor = child; }
    ctor.prototype = parent.prototype;
    child.prototype = new ctor();
  }

  function peg$SyntaxError(message, expected, found, location) {
    this.message  = message;
    this.expected = expected;
    this.found    = found;
    this.location = location;
    this.name     = "SyntaxError";

    if (typeof Error.captureStackTrace === "function") {
      Error.captureStackTrace(this, peg$SyntaxError);
    }
  }

  peg$subclass(peg$SyntaxError, Error);

  function peg$parse(input) {
    var options = arguments.length > 1 ? arguments[1] : {},
        parser  = this,

        peg$FAILED = {},

        peg$startRuleFunctions = { start: peg$parsestart },
        peg$startRuleFunction  = peg$parsestart,

        peg$c0 = /^[\n]/,
        peg$c1 = { type: "class", value: "[\\n]", description: "[\\n]" },
        peg$c2 = function(ingredient, amount, container, unit) {
            var result = {
              amount: amount,
              container: container,
              ingredient: ingredient,
              unit: unit
            };

            for(var i in result) {
              if(result[i] === null || result[i] === undefined) {
                delete result[i];
              }
            }

            return result;
          },
        peg$c3 = function(amount, container, unit, ingredient) {
            var result = {
              amount: amount,
              container: container,
              ingredient: ingredient,
              unit: unit
            };

            for(var i in result) {
              if(result[i] === null || result[i] === undefined) {
                delete result[i];
              }
            }

            return result;
          },
        peg$c4 = "(",
        peg$c5 = { type: "literal", value: "(", description: "\"(\"" },
        peg$c6 = ")",
        peg$c7 = { type: "literal", value: ")", description: "\")\"" },
        peg$c8 = function(amount, unit) {
            return { amount: amount, unit: unit };
          },
        peg$c9 = " ",
        peg$c10 = { type: "literal", value: " ", description: "\" \"" },
        peg$c11 = /^[\t]/,
        peg$c12 = { type: "class", value: "[\\t]", description: "[\\t]" },
        peg$c13 = ": ",
        peg$c14 = { type: "literal", value: ": ", description: "\": \"" },
        peg$c15 = function(letters) { return letters.join(''); },
        peg$c16 = /^[.]/,
        peg$c17 = { type: "class", value: "[.]", description: "[.]" },
        peg$c18 = /^[\/]/,
        peg$c19 = { type: "class", value: "[/]", description: "[/]" },
        peg$c20 = /^[0-9]/,
        peg$c21 = { type: "class", value: "[0-9]", description: "[0-9]" },
        peg$c22 = function(digits) { return digits.join(''); },
        peg$c23 = function(integer) {
            return integer;
          },
        peg$c24 = /^[a-zA-Z]/,
        peg$c25 = { type: "class", value: "[a-zA-Z]", description: "[a-zA-Z]" },
        peg$c26 = /^[a-zA-Z0-9\xC4\xE4\xDC\xFC\xD6\xDF\xF6'",.;()\-_%&*$+=]/,
        peg$c27 = { type: "class", value: "[a-zA-Z0-9\\xC4\\xE4\\xDC\\xFC\\xD6\\xDF\\xF6'\",.;\\(\\)\\-_%\\&\\*\\$\\+\\=]", description: "[a-zA-Z0-9\\xC4\\xE4\\xDC\\xFC\\xD6\\xDF\\xF6'\",.;\\(\\)\\-_%\\&\\*\\$\\+\\=]" },
        peg$c28 = "cups",
        peg$c29 = { type: "literal", value: "cups", description: "\"cups\"" },
        peg$c30 = "cup",
        peg$c31 = { type: "literal", value: "cup", description: "\"cup\"" },
        peg$c32 = "c.",
        peg$c33 = { type: "literal", value: "c.", description: "\"c.\"" },
        peg$c34 = "c",
        peg$c35 = { type: "literal", value: "c", description: "\"c\"" },
        peg$c36 = "tassen",
        peg$c37 = { type: "literal", value: "tassen", description: "\"tassen\"" },
        peg$c38 = "tasse",
        peg$c39 = { type: "literal", value: "tasse", description: "\"tasse\"" },
        peg$c40 = "tas.",
        peg$c41 = { type: "literal", value: "tas.", description: "\"tas.\"" },
        peg$c42 = "t.",
        peg$c43 = { type: "literal", value: "t.", description: "\"t.\"" },
        peg$c44 = "fluid",
        peg$c45 = { type: "literal", value: "fluid", description: "\"fluid\"" },
        peg$c46 = "fl",
        peg$c47 = { type: "literal", value: "fl", description: "\"fl\"" },
        peg$c48 = ".",
        peg$c49 = { type: "literal", value: ".", description: "\".\"" },
        peg$c50 = "gallons",
        peg$c51 = { type: "literal", value: "gallons", description: "\"gallons\"" },
        peg$c52 = "gallon",
        peg$c53 = { type: "literal", value: "gallon", description: "\"gallon\"" },
        peg$c54 = "gal.",
        peg$c55 = { type: "literal", value: "gal.", description: "\"gal.\"" },
        peg$c56 = "gal",
        peg$c57 = { type: "literal", value: "gal", description: "\"gal\"" },
        peg$c58 = "ounces",
        peg$c59 = { type: "literal", value: "ounces", description: "\"ounces\"" },
        peg$c60 = "ounce",
        peg$c61 = { type: "literal", value: "ounce", description: "\"ounce\"" },
        peg$c62 = "oz.",
        peg$c63 = { type: "literal", value: "oz.", description: "\"oz.\"" },
        peg$c64 = "oz",
        peg$c65 = { type: "literal", value: "oz", description: "\"oz\"" },
        peg$c66 = "pints",
        peg$c67 = { type: "literal", value: "pints", description: "\"pints\"" },
        peg$c68 = "pint",
        peg$c69 = { type: "literal", value: "pint", description: "\"pint\"" },
        peg$c70 = "pt.",
        peg$c71 = { type: "literal", value: "pt.", description: "\"pt.\"" },
        peg$c72 = "pt",
        peg$c73 = { type: "literal", value: "pt", description: "\"pt\"" },
        peg$c74 = "pounds",
        peg$c75 = { type: "literal", value: "pounds", description: "\"pounds\"" },
        peg$c76 = "pound",
        peg$c77 = { type: "literal", value: "pound", description: "\"pound\"" },
        peg$c78 = "lbs.",
        peg$c79 = { type: "literal", value: "lbs.", description: "\"lbs.\"" },
        peg$c80 = "lbs",
        peg$c81 = { type: "literal", value: "lbs", description: "\"lbs\"" },
        peg$c82 = "lb.",
        peg$c83 = { type: "literal", value: "lb.", description: "\"lb.\"" },
        peg$c84 = "lb",
        peg$c85 = { type: "literal", value: "lb", description: "\"lb\"" },
        peg$c86 = "pfunde",
        peg$c87 = { type: "literal", value: "pfunde", description: "\"pfunde\"" },
        peg$c88 = "pfund",
        peg$c89 = { type: "literal", value: "pfund", description: "\"pfund\"" },
        peg$c90 = "quarts",
        peg$c91 = { type: "literal", value: "quarts", description: "\"quarts\"" },
        peg$c92 = "quart",
        peg$c93 = { type: "literal", value: "quart", description: "\"quart\"" },
        peg$c94 = "qts.",
        peg$c95 = { type: "literal", value: "qts.", description: "\"qts.\"" },
        peg$c96 = "qts",
        peg$c97 = { type: "literal", value: "qts", description: "\"qts\"" },
        peg$c98 = "qt.",
        peg$c99 = { type: "literal", value: "qt.", description: "\"qt.\"" },
        peg$c100 = "qt",
        peg$c101 = { type: "literal", value: "qt", description: "\"qt\"" },
        peg$c102 = "tablespoons",
        peg$c103 = { type: "literal", value: "tablespoons", description: "\"tablespoons\"" },
        peg$c104 = "tablespoon",
        peg$c105 = { type: "literal", value: "tablespoon", description: "\"tablespoon\"" },
        peg$c106 = "tbsp.",
        peg$c107 = { type: "literal", value: "tbsp.", description: "\"tbsp.\"" },
        peg$c108 = "tbsp",
        peg$c109 = { type: "literal", value: "tbsp", description: "\"tbsp\"" },
        peg$c110 = "tbs.",
        peg$c111 = { type: "literal", value: "tbs.", description: "\"tbs.\"" },
        peg$c112 = "tbs",
        peg$c113 = { type: "literal", value: "tbs", description: "\"tbs\"" },
        peg$c114 = "T.",
        peg$c115 = { type: "literal", value: "T.", description: "\"T.\"" },
        peg$c116 = "T",
        peg$c117 = { type: "literal", value: "T", description: "\"T\"" },
        peg$c118 = "geh\xE4ufter essl\xF6ffel",
        peg$c119 = { type: "literal", value: "geh\xE4ufter essl\xF6ffel", description: "\"geh\\xE4ufter essl\\xF6ffel\"" },
        peg$c120 = "geh\xE4ufter e\xDFl\xF6ffel",
        peg$c121 = { type: "literal", value: "geh\xE4ufter e\xDFl\xF6ffel", description: "\"geh\\xE4ufter e\\xDFl\\xF6ffel\"" },
        peg$c122 = "geh\xE4ufte essl\xF6ffel",
        peg$c123 = { type: "literal", value: "geh\xE4ufte essl\xF6ffel", description: "\"geh\\xE4ufte essl\\xF6ffel\"" },
        peg$c124 = "geh\xE4ufte e\xDFl\xF6ffel",
        peg$c125 = { type: "literal", value: "geh\xE4ufte e\xDFl\xF6ffel", description: "\"geh\\xE4ufte e\\xDFl\\xF6ffel\"" },
        peg$c126 = "geh\xE4ufter el",
        peg$c127 = { type: "literal", value: "geh\xE4ufter el", description: "\"geh\\xE4ufter el\"" },
        peg$c128 = "geh\xE4ufte el",
        peg$c129 = { type: "literal", value: "geh\xE4ufte el", description: "\"geh\\xE4ufte el\"" },
        peg$c130 = "gestrichener essl\xF6ffel",
        peg$c131 = { type: "literal", value: "gestrichener essl\xF6ffel", description: "\"gestrichener essl\\xF6ffel\"" },
        peg$c132 = "gestrichener e\xDFl\xF6ffel",
        peg$c133 = { type: "literal", value: "gestrichener e\xDFl\xF6ffel", description: "\"gestrichener e\\xDFl\\xF6ffel\"" },
        peg$c134 = "gestrichene essl\xF6ffel",
        peg$c135 = { type: "literal", value: "gestrichene essl\xF6ffel", description: "\"gestrichene essl\\xF6ffel\"" },
        peg$c136 = "gestrichene e\xDFl\xF6ffel",
        peg$c137 = { type: "literal", value: "gestrichene e\xDFl\xF6ffel", description: "\"gestrichene e\\xDFl\\xF6ffel\"" },
        peg$c138 = "gestr. essl\xF6ffel",
        peg$c139 = { type: "literal", value: "gestr. essl\xF6ffel", description: "\"gestr. essl\\xF6ffel\"" },
        peg$c140 = "gestr. e\xDFl\xF6ffel",
        peg$c141 = { type: "literal", value: "gestr. e\xDFl\xF6ffel", description: "\"gestr. e\\xDFl\\xF6ffel\"" },
        peg$c142 = "gestr. el",
        peg$c143 = { type: "literal", value: "gestr. el", description: "\"gestr. el\"" },
        peg$c144 = "gestr el",
        peg$c145 = { type: "literal", value: "gestr el", description: "\"gestr el\"" },
        peg$c146 = "essl\xF6ffel",
        peg$c147 = { type: "literal", value: "essl\xF6ffel", description: "\"essl\\xF6ffel\"" },
        peg$c148 = "e\xDFl\xF6ffel",
        peg$c149 = { type: "literal", value: "e\xDFl\xF6ffel", description: "\"e\\xDFl\\xF6ffel\"" },
        peg$c150 = "esslfl.",
        peg$c151 = { type: "literal", value: "esslfl.", description: "\"esslfl.\"" },
        peg$c152 = "esslfl",
        peg$c153 = { type: "literal", value: "esslfl", description: "\"esslfl\"" },
        peg$c154 = "el",
        peg$c155 = { type: "literal", value: "el", description: "\"el\"" },
        peg$c156 = "teaspoons",
        peg$c157 = { type: "literal", value: "teaspoons", description: "\"teaspoons\"" },
        peg$c158 = "teaspoon",
        peg$c159 = { type: "literal", value: "teaspoon", description: "\"teaspoon\"" },
        peg$c160 = "tsp.",
        peg$c161 = { type: "literal", value: "tsp.", description: "\"tsp.\"" },
        peg$c162 = "tsp",
        peg$c163 = { type: "literal", value: "tsp", description: "\"tsp\"" },
        peg$c164 = "t",
        peg$c165 = { type: "literal", value: "t", description: "\"t\"" },
        peg$c166 = "geh\xE4ufter teel\xF6ffel",
        peg$c167 = { type: "literal", value: "geh\xE4ufter teel\xF6ffel", description: "\"geh\\xE4ufter teel\\xF6ffel\"" },
        peg$c168 = "geh\xE4ufte teel\xF6ffel",
        peg$c169 = { type: "literal", value: "geh\xE4ufte teel\xF6ffel", description: "\"geh\\xE4ufte teel\\xF6ffel\"" },
        peg$c170 = "geh\xE4ufte tl",
        peg$c171 = { type: "literal", value: "geh\xE4ufte tl", description: "\"geh\\xE4ufte tl\"" },
        peg$c172 = "geh\xE4ufter tl",
        peg$c173 = { type: "literal", value: "geh\xE4ufter tl", description: "\"geh\\xE4ufter tl\"" },
        peg$c174 = "gestrichener teel\xF6ffel",
        peg$c175 = { type: "literal", value: "gestrichener teel\xF6ffel", description: "\"gestrichener teel\\xF6ffel\"" },
        peg$c176 = "gestrichener tl",
        peg$c177 = { type: "literal", value: "gestrichener tl", description: "\"gestrichener tl\"" },
        peg$c178 = "gestrichene teel\xF6ffel",
        peg$c179 = { type: "literal", value: "gestrichene teel\xF6ffel", description: "\"gestrichene teel\\xF6ffel\"" },
        peg$c180 = "gestrichene tl",
        peg$c181 = { type: "literal", value: "gestrichene tl", description: "\"gestrichene tl\"" },
        peg$c182 = "gestr. teel\xF6ffel",
        peg$c183 = { type: "literal", value: "gestr. teel\xF6ffel", description: "\"gestr. teel\\xF6ffel\"" },
        peg$c184 = "gestr. tl",
        peg$c185 = { type: "literal", value: "gestr. tl", description: "\"gestr. tl\"" },
        peg$c186 = "teel\xF6ffel",
        peg$c187 = { type: "literal", value: "teel\xF6ffel", description: "\"teel\\xF6ffel\"" },
        peg$c188 = "teelfl.",
        peg$c189 = { type: "literal", value: "teelfl.", description: "\"teelfl.\"" },
        peg$c190 = "teelfl",
        peg$c191 = { type: "literal", value: "teelfl", description: "\"teelfl\"" },
        peg$c192 = "tl",
        peg$c193 = { type: "literal", value: "tl", description: "\"tl\"" },
        peg$c194 = "messerspitzen",
        peg$c195 = { type: "literal", value: "messerspitzen", description: "\"messerspitzen\"" },
        peg$c196 = "messerspitze",
        peg$c197 = { type: "literal", value: "messerspitze", description: "\"messerspitze\"" },
        peg$c198 = "msp.",
        peg$c199 = { type: "literal", value: "msp.", description: "\"msp.\"" },
        peg$c200 = "msp",
        peg$c201 = { type: "literal", value: "msp", description: "\"msp\"" },
        peg$c202 = "prisen",
        peg$c203 = { type: "literal", value: "prisen", description: "\"prisen\"" },
        peg$c204 = "prise",
        peg$c205 = { type: "literal", value: "prise", description: "\"prise\"" },
        peg$c206 = "pr.",
        peg$c207 = { type: "literal", value: "pr.", description: "\"pr.\"" },
        peg$c208 = "pr",
        peg$c209 = { type: "literal", value: "pr", description: "\"pr\"" },
        peg$c210 = "sch\xFCsse",
        peg$c211 = { type: "literal", value: "sch\xFCsse", description: "\"sch\\xFCsse\"" },
        peg$c212 = "schuss",
        peg$c213 = { type: "literal", value: "schuss", description: "\"schuss\"" },
        peg$c214 = "spritzer",
        peg$c215 = { type: "literal", value: "spritzer", description: "\"spritzer\"" },
        peg$c216 = "tropfen",
        peg$c217 = { type: "literal", value: "tropfen", description: "\"tropfen\"" },
        peg$c218 = "b\xFCnde",
        peg$c219 = { type: "literal", value: "b\xFCnde", description: "\"b\\xFCnde\"" },
        peg$c220 = "bunde",
        peg$c221 = { type: "literal", value: "bunde", description: "\"bunde\"" },
        peg$c222 = "bund",
        peg$c223 = { type: "literal", value: "bund", description: "\"bund\"" },
        peg$c224 = "grams",
        peg$c225 = { type: "literal", value: "grams", description: "\"grams\"" },
        peg$c226 = "gram",
        peg$c227 = { type: "literal", value: "gram", description: "\"gram\"" },
        peg$c228 = "gr.",
        peg$c229 = { type: "literal", value: "gr.", description: "\"gr.\"" },
        peg$c230 = "gr",
        peg$c231 = { type: "literal", value: "gr", description: "\"gr\"" },
        peg$c232 = "g.",
        peg$c233 = { type: "literal", value: "g.", description: "\"g.\"" },
        peg$c234 = "g",
        peg$c235 = { type: "literal", value: "g", description: "\"g\"" },
        peg$c236 = "kilograms",
        peg$c237 = { type: "literal", value: "kilograms", description: "\"kilograms\"" },
        peg$c238 = "kilogram",
        peg$c239 = { type: "literal", value: "kilogram", description: "\"kilogram\"" },
        peg$c240 = "kg.",
        peg$c241 = { type: "literal", value: "kg.", description: "\"kg.\"" },
        peg$c242 = "kg",
        peg$c243 = { type: "literal", value: "kg", description: "\"kg\"" },
        peg$c244 = "liters",
        peg$c245 = { type: "literal", value: "liters", description: "\"liters\"" },
        peg$c246 = "liter",
        peg$c247 = { type: "literal", value: "liter", description: "\"liter\"" },
        peg$c248 = "l.",
        peg$c249 = { type: "literal", value: "l.", description: "\"l.\"" },
        peg$c250 = "l",
        peg$c251 = { type: "literal", value: "l", description: "\"l\"" },
        peg$c252 = "milligrams",
        peg$c253 = { type: "literal", value: "milligrams", description: "\"milligrams\"" },
        peg$c254 = "milligram",
        peg$c255 = { type: "literal", value: "milligram", description: "\"milligram\"" },
        peg$c256 = "mg.",
        peg$c257 = { type: "literal", value: "mg.", description: "\"mg.\"" },
        peg$c258 = "mg",
        peg$c259 = { type: "literal", value: "mg", description: "\"mg\"" },
        peg$c260 = "milliliters",
        peg$c261 = { type: "literal", value: "milliliters", description: "\"milliliters\"" },
        peg$c262 = "milliliter",
        peg$c263 = { type: "literal", value: "milliliter", description: "\"milliliter\"" },
        peg$c264 = "ml.",
        peg$c265 = { type: "literal", value: "ml.", description: "\"ml.\"" },
        peg$c266 = "ml",
        peg$c267 = { type: "literal", value: "ml", description: "\"ml\"" },
        peg$c268 = "dashes",
        peg$c269 = { type: "literal", value: "dashes", description: "\"dashes\"" },
        peg$c270 = "dash",
        peg$c271 = { type: "literal", value: "dash", description: "\"dash\"" },
        peg$c272 = "handfuls",
        peg$c273 = { type: "literal", value: "handfuls", description: "\"handfuls\"" },
        peg$c274 = "handful",
        peg$c275 = { type: "literal", value: "handful", description: "\"handful\"" },
        peg$c276 = "pinches",
        peg$c277 = { type: "literal", value: "pinches", description: "\"pinches\"" },
        peg$c278 = "pinch",
        peg$c279 = { type: "literal", value: "pinch", description: "\"pinch\"" },
        peg$c280 = "touches",
        peg$c281 = { type: "literal", value: "touches", description: "\"touches\"" },
        peg$c282 = "touch",
        peg$c283 = { type: "literal", value: "touch", description: "\"touch\"" },
        peg$c284 = "etwas",
        peg$c285 = { type: "literal", value: "etwas", description: "\"etwas\"" },

        peg$currPos          = 0,
        peg$savedPos         = 0,
        peg$posDetailsCache  = [{ line: 1, column: 1, seenCR: false }],
        peg$maxFailPos       = 0,
        peg$maxFailExpected  = [],
        peg$silentFails      = 0,

        peg$result;

    if ("startRule" in options) {
      if (!(options.startRule in peg$startRuleFunctions)) {
        throw new Error("Can't start parsing from rule \"" + options.startRule + "\".");
      }

      peg$startRuleFunction = peg$startRuleFunctions[options.startRule];
    }

    function text() {
      return input.substring(peg$savedPos, peg$currPos);
    }

    function location() {
      return peg$computeLocation(peg$savedPos, peg$currPos);
    }

    function expected(description) {
      throw peg$buildException(
        null,
        [{ type: "other", description: description }],
        input.substring(peg$savedPos, peg$currPos),
        peg$computeLocation(peg$savedPos, peg$currPos)
      );
    }

    function error(message) {
      throw peg$buildException(
        message,
        null,
        input.substring(peg$savedPos, peg$currPos),
        peg$computeLocation(peg$savedPos, peg$currPos)
      );
    }

    function peg$computePosDetails(pos) {
      var details = peg$posDetailsCache[pos],
          p, ch;

      if (details) {
        return details;
      } else {
        p = pos - 1;
        while (!peg$posDetailsCache[p]) {
          p--;
        }

        details = peg$posDetailsCache[p];
        details = {
          line:   details.line,
          column: details.column,
          seenCR: details.seenCR
        };

        while (p < pos) {
          ch = input.charAt(p);
          if (ch === "\n") {
            if (!details.seenCR) { details.line++; }
            details.column = 1;
            details.seenCR = false;
          } else if (ch === "\r" || ch === "\u2028" || ch === "\u2029") {
            details.line++;
            details.column = 1;
            details.seenCR = true;
          } else {
            details.column++;
            details.seenCR = false;
          }

          p++;
        }

        peg$posDetailsCache[pos] = details;
        return details;
      }
    }

    function peg$computeLocation(startPos, endPos) {
      var startPosDetails = peg$computePosDetails(startPos),
          endPosDetails   = peg$computePosDetails(endPos);

      return {
        start: {
          offset: startPos,
          line:   startPosDetails.line,
          column: startPosDetails.column
        },
        end: {
          offset: endPos,
          line:   endPosDetails.line,
          column: endPosDetails.column
        }
      };
    }

    function peg$fail(expected) {
      if (peg$currPos < peg$maxFailPos) { return; }

      if (peg$currPos > peg$maxFailPos) {
        peg$maxFailPos = peg$currPos;
        peg$maxFailExpected = [];
      }

      peg$maxFailExpected.push(expected);
    }

    function peg$buildException(message, expected, found, location) {
      function cleanupExpected(expected) {
        var i = 1;

        expected.sort(function(a, b) {
          if (a.description < b.description) {
            return -1;
          } else if (a.description > b.description) {
            return 1;
          } else {
            return 0;
          }
        });

        while (i < expected.length) {
          if (expected[i - 1] === expected[i]) {
            expected.splice(i, 1);
          } else {
            i++;
          }
        }
      }

      function buildMessage(expected, found) {
        function stringEscape(s) {
          function hex(ch) { return ch.charCodeAt(0).toString(16).toUpperCase(); }

          return s
            .replace(/\\/g,   '\\\\')
            .replace(/"/g,    '\\"')
            .replace(/\x08/g, '\\b')
            .replace(/\t/g,   '\\t')
            .replace(/\n/g,   '\\n')
            .replace(/\f/g,   '\\f')
            .replace(/\r/g,   '\\r')
            .replace(/[\x00-\x07\x0B\x0E\x0F]/g, function(ch) { return '\\x0' + hex(ch); })
            .replace(/[\x10-\x1F\x80-\xFF]/g,    function(ch) { return '\\x'  + hex(ch); })
            .replace(/[\u0100-\u0FFF]/g,         function(ch) { return '\\u0' + hex(ch); })
            .replace(/[\u1000-\uFFFF]/g,         function(ch) { return '\\u'  + hex(ch); });
        }

        var expectedDescs = new Array(expected.length),
            expectedDesc, foundDesc, i;

        for (i = 0; i < expected.length; i++) {
          expectedDescs[i] = expected[i].description;
        }

        expectedDesc = expected.length > 1
          ? expectedDescs.slice(0, -1).join(", ")
              + " or "
              + expectedDescs[expected.length - 1]
          : expectedDescs[0];

        foundDesc = found ? "\"" + stringEscape(found) + "\"" : "end of input";

        return "Expected " + expectedDesc + " but " + foundDesc + " found.";
      }

      if (expected !== null) {
        cleanupExpected(expected);
      }

      return new peg$SyntaxError(
        message !== null ? message : buildMessage(expected, found),
        expected,
        found,
        location
      );
    }

    function peg$parsestart() {
      var s0;

      s0 = peg$parseingredient_addition();

      return s0;
    }

    function peg$parseingredient_addition() {
      var s0, s1, s2, s3, s4, s5, s6, s7, s8;

      s0 = peg$currPos;
      s1 = peg$parsephrase();
      if (s1 === peg$FAILED) {
        s1 = null;
      }
      if (s1 !== peg$FAILED) {
        s2 = peg$parsecolon();
        if (s2 !== peg$FAILED) {
          s3 = peg$parseamount();
          if (s3 === peg$FAILED) {
            s3 = null;
          }
          if (s3 !== peg$FAILED) {
            s4 = [];
            s5 = peg$parsews();
            if (s5 !== peg$FAILED) {
              while (s5 !== peg$FAILED) {
                s4.push(s5);
                s5 = peg$parsews();
              }
            } else {
              s4 = peg$FAILED;
            }
            if (s4 === peg$FAILED) {
              s4 = null;
            }
            if (s4 !== peg$FAILED) {
              s5 = peg$parsecontainer();
              if (s5 === peg$FAILED) {
                s5 = null;
              }
              if (s5 !== peg$FAILED) {
                s6 = peg$parseunit();
                if (s6 === peg$FAILED) {
                  s6 = null;
                }
                if (s6 !== peg$FAILED) {
                  s7 = [];
                  s8 = peg$parsews();
                  if (s8 !== peg$FAILED) {
                    while (s8 !== peg$FAILED) {
                      s7.push(s8);
                      s8 = peg$parsews();
                    }
                  } else {
                    s7 = peg$FAILED;
                  }
                  if (s7 === peg$FAILED) {
                    s7 = null;
                  }
                  if (s7 !== peg$FAILED) {
                    if (peg$c0.test(input.charAt(peg$currPos))) {
                      s8 = input.charAt(peg$currPos);
                      peg$currPos++;
                    } else {
                      s8 = peg$FAILED;
                      if (peg$silentFails === 0) { peg$fail(peg$c1); }
                    }
                    if (s8 === peg$FAILED) {
                      s8 = null;
                    }
                    if (s8 !== peg$FAILED) {
                      peg$savedPos = s0;
                      s1 = peg$c2(s1, s3, s5, s6);
                      s0 = s1;
                    } else {
                      peg$currPos = s0;
                      s0 = peg$FAILED;
                    }
                  } else {
                    peg$currPos = s0;
                    s0 = peg$FAILED;
                  }
                } else {
                  peg$currPos = s0;
                  s0 = peg$FAILED;
                }
              } else {
                peg$currPos = s0;
                s0 = peg$FAILED;
              }
            } else {
              peg$currPos = s0;
              s0 = peg$FAILED;
            }
          } else {
            peg$currPos = s0;
            s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
      if (s0 === peg$FAILED) {
        s0 = peg$currPos;
        s1 = peg$parseamount();
        if (s1 === peg$FAILED) {
          s1 = null;
        }
        if (s1 !== peg$FAILED) {
          s2 = [];
          s3 = peg$parsews();
          if (s3 !== peg$FAILED) {
            while (s3 !== peg$FAILED) {
              s2.push(s3);
              s3 = peg$parsews();
            }
          } else {
            s2 = peg$FAILED;
          }
          if (s2 === peg$FAILED) {
            s2 = null;
          }
          if (s2 !== peg$FAILED) {
            s3 = peg$parsecontainer();
            if (s3 === peg$FAILED) {
              s3 = null;
            }
            if (s3 !== peg$FAILED) {
              s4 = peg$parseunit();
              if (s4 === peg$FAILED) {
                s4 = null;
              }
              if (s4 !== peg$FAILED) {
                s5 = [];
                s6 = peg$parsews();
                if (s6 !== peg$FAILED) {
                  while (s6 !== peg$FAILED) {
                    s5.push(s6);
                    s6 = peg$parsews();
                  }
                } else {
                  s5 = peg$FAILED;
                }
                if (s5 === peg$FAILED) {
                  s5 = null;
                }
                if (s5 !== peg$FAILED) {
                  s6 = peg$parsephrase();
                  if (s6 === peg$FAILED) {
                    s6 = null;
                  }
                  if (s6 !== peg$FAILED) {
                    if (peg$c0.test(input.charAt(peg$currPos))) {
                      s7 = input.charAt(peg$currPos);
                      peg$currPos++;
                    } else {
                      s7 = peg$FAILED;
                      if (peg$silentFails === 0) { peg$fail(peg$c1); }
                    }
                    if (s7 === peg$FAILED) {
                      s7 = null;
                    }
                    if (s7 !== peg$FAILED) {
                      peg$savedPos = s0;
                      s1 = peg$c3(s1, s3, s4, s6);
                      s0 = s1;
                    } else {
                      peg$currPos = s0;
                      s0 = peg$FAILED;
                    }
                  } else {
                    peg$currPos = s0;
                    s0 = peg$FAILED;
                  }
                } else {
                  peg$currPos = s0;
                  s0 = peg$FAILED;
                }
              } else {
                peg$currPos = s0;
                s0 = peg$FAILED;
              }
            } else {
              peg$currPos = s0;
              s0 = peg$FAILED;
            }
          } else {
            peg$currPos = s0;
            s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      }

      return s0;
    }

    function peg$parseamount() {
      var s0;

      s0 = peg$parsefloat();
      if (s0 === peg$FAILED) {
        s0 = peg$parsemixed_number();
        if (s0 === peg$FAILED) {
          s0 = peg$parsefraction();
          if (s0 === peg$FAILED) {
            s0 = peg$parseinteger();
          }
        }
      }

      return s0;
    }

    function peg$parsecontainer() {
      var s0, s1, s2, s3, s4, s5;

      s0 = peg$currPos;
      if (input.charCodeAt(peg$currPos) === 40) {
        s1 = peg$c4;
        peg$currPos++;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c5); }
      }
      if (s1 === peg$FAILED) {
        s1 = null;
      }
      if (s1 !== peg$FAILED) {
        s2 = peg$parseamount();
        if (s2 !== peg$FAILED) {
          s3 = [];
          s4 = peg$parsews();
          if (s4 !== peg$FAILED) {
            while (s4 !== peg$FAILED) {
              s3.push(s4);
              s4 = peg$parsews();
            }
          } else {
            s3 = peg$FAILED;
          }
          if (s3 === peg$FAILED) {
            s3 = null;
          }
          if (s3 !== peg$FAILED) {
            s4 = peg$parseunit();
            if (s4 !== peg$FAILED) {
              if (input.charCodeAt(peg$currPos) === 41) {
                s5 = peg$c6;
                peg$currPos++;
              } else {
                s5 = peg$FAILED;
                if (peg$silentFails === 0) { peg$fail(peg$c7); }
              }
              if (s5 === peg$FAILED) {
                s5 = null;
              }
              if (s5 !== peg$FAILED) {
                peg$savedPos = s0;
                s1 = peg$c8(s2, s4);
                s0 = s1;
              } else {
                peg$currPos = s0;
                s0 = peg$FAILED;
              }
            } else {
              peg$currPos = s0;
              s0 = peg$FAILED;
            }
          } else {
            peg$currPos = s0;
            s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }

      return s0;
    }

    function peg$parsews() {
      var s0;

      if (input.charCodeAt(peg$currPos) === 32) {
        s0 = peg$c9;
        peg$currPos++;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c10); }
      }
      if (s0 === peg$FAILED) {
        if (peg$c11.test(input.charAt(peg$currPos))) {
          s0 = input.charAt(peg$currPos);
          peg$currPos++;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c12); }
        }
      }

      return s0;
    }

    function peg$parsecolon() {
      var s0;

      if (input.substr(peg$currPos, 2) === peg$c13) {
        s0 = peg$c13;
        peg$currPos += 2;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c14); }
      }

      return s0;
    }

    function peg$parsespace() {
      var s0;

      if (input.charCodeAt(peg$currPos) === 32) {
        s0 = peg$c9;
        peg$currPos++;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c10); }
      }

      return s0;
    }

    function peg$parsephrase() {
      var s0, s1, s2, s3, s4, s5, s6, s7;

      s0 = peg$currPos;
      s1 = peg$currPos;
      s2 = peg$parseword();
      if (s2 !== peg$FAILED) {
        s3 = [];
        s4 = peg$currPos;
        s5 = peg$parsespace();
        if (s5 !== peg$FAILED) {
          s6 = [];
          s7 = peg$parseword();
          while (s7 !== peg$FAILED) {
            s6.push(s7);
            s7 = peg$parseword();
          }
          if (s6 !== peg$FAILED) {
            s5 = [s5, s6];
            s4 = s5;
          } else {
            peg$currPos = s4;
            s4 = peg$FAILED;
          }
        } else {
          peg$currPos = s4;
          s4 = peg$FAILED;
        }
        while (s4 !== peg$FAILED) {
          s3.push(s4);
          s4 = peg$currPos;
          s5 = peg$parsespace();
          if (s5 !== peg$FAILED) {
            s6 = [];
            s7 = peg$parseword();
            while (s7 !== peg$FAILED) {
              s6.push(s7);
              s7 = peg$parseword();
            }
            if (s6 !== peg$FAILED) {
              s5 = [s5, s6];
              s4 = s5;
            } else {
              peg$currPos = s4;
              s4 = peg$FAILED;
            }
          } else {
            peg$currPos = s4;
            s4 = peg$FAILED;
          }
        }
        if (s3 !== peg$FAILED) {
          s2 = [s2, s3];
          s1 = s2;
        } else {
          peg$currPos = s1;
          s1 = peg$FAILED;
        }
      } else {
        peg$currPos = s1;
        s1 = peg$FAILED;
      }
      if (s1 !== peg$FAILED) {
        s0 = input.substring(s0, peg$currPos);
      } else {
        s0 = s1;
      }

      return s0;
    }

    function peg$parseword() {
      var s0, s1, s2;

      s0 = peg$currPos;
      s1 = [];
      s2 = peg$parsechar();
      if (s2 !== peg$FAILED) {
        while (s2 !== peg$FAILED) {
          s1.push(s2);
          s2 = peg$parsechar();
        }
      } else {
        s1 = peg$FAILED;
      }
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$c15(s1);
      }
      s0 = s1;

      return s0;
    }

    function peg$parsefloat() {
      var s0, s1, s2, s3, s4;

      s0 = peg$currPos;
      s1 = peg$currPos;
      s2 = peg$parseinteger();
      if (s2 === peg$FAILED) {
        s2 = null;
      }
      if (s2 !== peg$FAILED) {
        if (peg$c16.test(input.charAt(peg$currPos))) {
          s3 = input.charAt(peg$currPos);
          peg$currPos++;
        } else {
          s3 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c17); }
        }
        if (s3 !== peg$FAILED) {
          s4 = peg$parseinteger();
          if (s4 !== peg$FAILED) {
            s2 = [s2, s3, s4];
            s1 = s2;
          } else {
            peg$currPos = s1;
            s1 = peg$FAILED;
          }
        } else {
          peg$currPos = s1;
          s1 = peg$FAILED;
        }
      } else {
        peg$currPos = s1;
        s1 = peg$FAILED;
      }
      if (s1 !== peg$FAILED) {
        s0 = input.substring(s0, peg$currPos);
      } else {
        s0 = s1;
      }

      return s0;
    }

    function peg$parsemixed_number() {
      var s0, s1, s2, s3, s4;

      s0 = peg$currPos;
      s1 = peg$currPos;
      s2 = peg$parseinteger();
      if (s2 !== peg$FAILED) {
        s3 = peg$parsespace();
        if (s3 !== peg$FAILED) {
          s4 = peg$parsefraction();
          if (s4 !== peg$FAILED) {
            s2 = [s2, s3, s4];
            s1 = s2;
          } else {
            peg$currPos = s1;
            s1 = peg$FAILED;
          }
        } else {
          peg$currPos = s1;
          s1 = peg$FAILED;
        }
      } else {
        peg$currPos = s1;
        s1 = peg$FAILED;
      }
      if (s1 !== peg$FAILED) {
        s0 = input.substring(s0, peg$currPos);
      } else {
        s0 = s1;
      }

      return s0;
    }

    function peg$parsefraction() {
      var s0, s1, s2, s3, s4;

      s0 = peg$currPos;
      s1 = peg$currPos;
      s2 = peg$parseinteger();
      if (s2 !== peg$FAILED) {
        if (peg$c18.test(input.charAt(peg$currPos))) {
          s3 = input.charAt(peg$currPos);
          peg$currPos++;
        } else {
          s3 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c19); }
        }
        if (s3 !== peg$FAILED) {
          s4 = peg$parseinteger();
          if (s4 !== peg$FAILED) {
            s2 = [s2, s3, s4];
            s1 = s2;
          } else {
            peg$currPos = s1;
            s1 = peg$FAILED;
          }
        } else {
          peg$currPos = s1;
          s1 = peg$FAILED;
        }
      } else {
        peg$currPos = s1;
        s1 = peg$FAILED;
      }
      if (s1 !== peg$FAILED) {
        s0 = input.substring(s0, peg$currPos);
      } else {
        s0 = s1;
      }

      return s0;
    }

    function peg$parseinteger() {
      var s0, s1, s2;

      s0 = peg$currPos;
      s1 = [];
      if (peg$c20.test(input.charAt(peg$currPos))) {
        s2 = input.charAt(peg$currPos);
        peg$currPos++;
      } else {
        s2 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c21); }
      }
      if (s2 !== peg$FAILED) {
        while (s2 !== peg$FAILED) {
          s1.push(s2);
          if (peg$c20.test(input.charAt(peg$currPos))) {
            s2 = input.charAt(peg$currPos);
            peg$currPos++;
          } else {
            s2 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c21); }
          }
        }
      } else {
        s1 = peg$FAILED;
      }
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$c22(s1);
      }
      s0 = s1;

      return s0;
    }

    function peg$parseparenthetical() {
      var s0, s1, s2, s3, s4, s5;

      s0 = peg$currPos;
      if (input.charCodeAt(peg$currPos) === 40) {
        s1 = peg$c4;
        peg$currPos++;
      } else {
        s1 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c5); }
      }
      if (s1 !== peg$FAILED) {
        s2 = [];
        s3 = peg$parsews();
        while (s3 !== peg$FAILED) {
          s2.push(s3);
          s3 = peg$parsews();
        }
        if (s2 !== peg$FAILED) {
          s3 = peg$parseinteger();
          if (s3 !== peg$FAILED) {
            s4 = [];
            s5 = peg$parsews();
            while (s5 !== peg$FAILED) {
              s4.push(s5);
              s5 = peg$parsews();
            }
            if (s4 !== peg$FAILED) {
              if (input.charCodeAt(peg$currPos) === 41) {
                s5 = peg$c6;
                peg$currPos++;
              } else {
                s5 = peg$FAILED;
                if (peg$silentFails === 0) { peg$fail(peg$c7); }
              }
              if (s5 !== peg$FAILED) {
                peg$savedPos = s0;
                s1 = peg$c23(s3);
                s0 = s1;
              } else {
                peg$currPos = s0;
                s0 = peg$FAILED;
              }
            } else {
              peg$currPos = s0;
              s0 = peg$FAILED;
            }
          } else {
            peg$currPos = s0;
            s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }

      return s0;
    }

    function peg$parseletter() {
      var s0;

      if (peg$c24.test(input.charAt(peg$currPos))) {
        s0 = input.charAt(peg$currPos);
        peg$currPos++;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c25); }
      }

      return s0;
    }

    function peg$parsechar() {
      var s0;

      if (peg$c26.test(input.charAt(peg$currPos))) {
        s0 = input.charAt(peg$currPos);
        peg$currPos++;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c27); }
      }

      return s0;
    }

    function peg$parseunit() {
      var s0, s1, s2, s3, s4;

      s0 = peg$currPos;
      s1 = peg$currPos;
      s2 = peg$parseenglish_unit();
      if (s2 !== peg$FAILED) {
        s3 = peg$currPos;
        peg$silentFails++;
        s4 = peg$parseletter();
        peg$silentFails--;
        if (s4 === peg$FAILED) {
          s3 = void 0;
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
        if (s3 !== peg$FAILED) {
          s2 = [s2, s3];
          s1 = s2;
        } else {
          peg$currPos = s1;
          s1 = peg$FAILED;
        }
      } else {
        peg$currPos = s1;
        s1 = peg$FAILED;
      }
      if (s1 !== peg$FAILED) {
        s0 = input.substring(s0, peg$currPos);
      } else {
        s0 = s1;
      }
      if (s0 === peg$FAILED) {
        s0 = peg$currPos;
        s1 = peg$currPos;
        s2 = peg$parsegerman_unit();
        if (s2 !== peg$FAILED) {
          s3 = peg$currPos;
          peg$silentFails++;
          s4 = peg$parseletter();
          peg$silentFails--;
          if (s4 === peg$FAILED) {
            s3 = void 0;
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
          if (s3 !== peg$FAILED) {
            s2 = [s2, s3];
            s1 = s2;
          } else {
            peg$currPos = s1;
            s1 = peg$FAILED;
          }
        } else {
          peg$currPos = s1;
          s1 = peg$FAILED;
        }
        if (s1 !== peg$FAILED) {
          s0 = input.substring(s0, peg$currPos);
        } else {
          s0 = s1;
        }
        if (s0 === peg$FAILED) {
          s0 = peg$currPos;
          s1 = peg$currPos;
          s2 = peg$parsemetric_unit();
          if (s2 !== peg$FAILED) {
            s3 = peg$currPos;
            peg$silentFails++;
            s4 = peg$parseletter();
            peg$silentFails--;
            if (s4 === peg$FAILED) {
              s3 = void 0;
            } else {
              peg$currPos = s3;
              s3 = peg$FAILED;
            }
            if (s3 !== peg$FAILED) {
              s2 = [s2, s3];
              s1 = s2;
            } else {
              peg$currPos = s1;
              s1 = peg$FAILED;
            }
          } else {
            peg$currPos = s1;
            s1 = peg$FAILED;
          }
          if (s1 !== peg$FAILED) {
            s0 = input.substring(s0, peg$currPos);
          } else {
            s0 = s1;
          }
          if (s0 === peg$FAILED) {
            s0 = peg$currPos;
            s1 = peg$currPos;
            s2 = peg$parseimprecise_unit();
            if (s2 !== peg$FAILED) {
              s3 = peg$currPos;
              peg$silentFails++;
              s4 = peg$parseletter();
              peg$silentFails--;
              if (s4 === peg$FAILED) {
                s3 = void 0;
              } else {
                peg$currPos = s3;
                s3 = peg$FAILED;
              }
              if (s3 !== peg$FAILED) {
                s2 = [s2, s3];
                s1 = s2;
              } else {
                peg$currPos = s1;
                s1 = peg$FAILED;
              }
            } else {
              peg$currPos = s1;
              s1 = peg$FAILED;
            }
            if (s1 !== peg$FAILED) {
              s0 = input.substring(s0, peg$currPos);
            } else {
              s0 = s1;
            }
          }
        }
      }

      return s0;
    }

    function peg$parseenglish_unit() {
      var s0;

      s0 = peg$parsecup();
      if (s0 === peg$FAILED) {
        s0 = peg$parsefluid_ounce();
        if (s0 === peg$FAILED) {
          s0 = peg$parsegallon();
          if (s0 === peg$FAILED) {
            s0 = peg$parseounce();
            if (s0 === peg$FAILED) {
              s0 = peg$parsepint();
              if (s0 === peg$FAILED) {
                s0 = peg$parsepound();
                if (s0 === peg$FAILED) {
                  s0 = peg$parsequart();
                  if (s0 === peg$FAILED) {
                    s0 = peg$parsetablespoon();
                    if (s0 === peg$FAILED) {
                      s0 = peg$parseteaspoon();
                    }
                  }
                }
              }
            }
          }
        }
      }

      return s0;
    }

    function peg$parsegerman_unit() {
      var s0;

      s0 = peg$parseTasse();
      if (s0 === peg$FAILED) {
        s0 = peg$parseEsslöffel();
        if (s0 === peg$FAILED) {
          s0 = peg$parseTeelöffel();
          if (s0 === peg$FAILED) {
            s0 = peg$parsePfund();
            if (s0 === peg$FAILED) {
              s0 = peg$parseMesserspitze();
              if (s0 === peg$FAILED) {
                s0 = peg$parsePrise();
                if (s0 === peg$FAILED) {
                  s0 = peg$parseSpritzer();
                  if (s0 === peg$FAILED) {
                    s0 = peg$parseSchuss();
                    if (s0 === peg$FAILED) {
                      s0 = peg$parseTropfen();
                      if (s0 === peg$FAILED) {
                        s0 = peg$parseBund();
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }

      return s0;
    }

    function peg$parsecup() {
      var s0;

      if (input.substr(peg$currPos, 4).toLowerCase() === peg$c28) {
        s0 = input.substr(peg$currPos, 4);
        peg$currPos += 4;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c29); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 3).toLowerCase() === peg$c30) {
          s0 = input.substr(peg$currPos, 3);
          peg$currPos += 3;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c31); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 2).toLowerCase() === peg$c32) {
            s0 = input.substr(peg$currPos, 2);
            peg$currPos += 2;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c33); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 1).toLowerCase() === peg$c34) {
              s0 = input.charAt(peg$currPos);
              peg$currPos++;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c35); }
            }
          }
        }
      }

      return s0;
    }

    function peg$parseTasse() {
      var s0;

      if (input.substr(peg$currPos, 6).toLowerCase() === peg$c36) {
        s0 = input.substr(peg$currPos, 6);
        peg$currPos += 6;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c37); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 5).toLowerCase() === peg$c38) {
          s0 = input.substr(peg$currPos, 5);
          peg$currPos += 5;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c39); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 4).toLowerCase() === peg$c40) {
            s0 = input.substr(peg$currPos, 4);
            peg$currPos += 4;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c41); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 2).toLowerCase() === peg$c42) {
              s0 = input.substr(peg$currPos, 2);
              peg$currPos += 2;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c43); }
            }
          }
        }
      }

      return s0;
    }

    function peg$parsefluid_ounce() {
      var s0, s1, s2, s3;

      s0 = peg$currPos;
      s1 = peg$parsefluid();
      if (s1 !== peg$FAILED) {
        s2 = peg$parsews();
        if (s2 !== peg$FAILED) {
          s3 = peg$parseounce();
          if (s3 !== peg$FAILED) {
            s1 = [s1, s2, s3];
            s0 = s1;
          } else {
            peg$currPos = s0;
            s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }

      return s0;
    }

    function peg$parsefluid() {
      var s0, s1, s2;

      if (input.substr(peg$currPos, 5).toLowerCase() === peg$c44) {
        s0 = input.substr(peg$currPos, 5);
        peg$currPos += 5;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c45); }
      }
      if (s0 === peg$FAILED) {
        s0 = peg$currPos;
        if (input.substr(peg$currPos, 2).toLowerCase() === peg$c46) {
          s1 = input.substr(peg$currPos, 2);
          peg$currPos += 2;
        } else {
          s1 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c47); }
        }
        if (s1 !== peg$FAILED) {
          if (input.charCodeAt(peg$currPos) === 46) {
            s2 = peg$c48;
            peg$currPos++;
          } else {
            s2 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c49); }
          }
          if (s2 === peg$FAILED) {
            s2 = null;
          }
          if (s2 !== peg$FAILED) {
            s1 = [s1, s2];
            s0 = s1;
          } else {
            peg$currPos = s0;
            s0 = peg$FAILED;
          }
        } else {
          peg$currPos = s0;
          s0 = peg$FAILED;
        }
      }

      return s0;
    }

    function peg$parsegallon() {
      var s0;

      if (input.substr(peg$currPos, 7).toLowerCase() === peg$c50) {
        s0 = input.substr(peg$currPos, 7);
        peg$currPos += 7;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c51); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 6).toLowerCase() === peg$c52) {
          s0 = input.substr(peg$currPos, 6);
          peg$currPos += 6;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c53); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 4).toLowerCase() === peg$c54) {
            s0 = input.substr(peg$currPos, 4);
            peg$currPos += 4;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c55); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 3).toLowerCase() === peg$c56) {
              s0 = input.substr(peg$currPos, 3);
              peg$currPos += 3;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c57); }
            }
          }
        }
      }

      return s0;
    }

    function peg$parseounce() {
      var s0;

      if (input.substr(peg$currPos, 6).toLowerCase() === peg$c58) {
        s0 = input.substr(peg$currPos, 6);
        peg$currPos += 6;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c59); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 5).toLowerCase() === peg$c60) {
          s0 = input.substr(peg$currPos, 5);
          peg$currPos += 5;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c61); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 3).toLowerCase() === peg$c62) {
            s0 = input.substr(peg$currPos, 3);
            peg$currPos += 3;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c63); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 2).toLowerCase() === peg$c64) {
              s0 = input.substr(peg$currPos, 2);
              peg$currPos += 2;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c65); }
            }
          }
        }
      }

      return s0;
    }

    function peg$parsepint() {
      var s0;

      if (input.substr(peg$currPos, 5).toLowerCase() === peg$c66) {
        s0 = input.substr(peg$currPos, 5);
        peg$currPos += 5;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c67); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 4).toLowerCase() === peg$c68) {
          s0 = input.substr(peg$currPos, 4);
          peg$currPos += 4;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c69); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 3).toLowerCase() === peg$c70) {
            s0 = input.substr(peg$currPos, 3);
            peg$currPos += 3;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c71); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 2).toLowerCase() === peg$c72) {
              s0 = input.substr(peg$currPos, 2);
              peg$currPos += 2;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c73); }
            }
          }
        }
      }

      return s0;
    }

    function peg$parsepound() {
      var s0;

      if (input.substr(peg$currPos, 6).toLowerCase() === peg$c74) {
        s0 = input.substr(peg$currPos, 6);
        peg$currPos += 6;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c75); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 5).toLowerCase() === peg$c76) {
          s0 = input.substr(peg$currPos, 5);
          peg$currPos += 5;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c77); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 4).toLowerCase() === peg$c78) {
            s0 = input.substr(peg$currPos, 4);
            peg$currPos += 4;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c79); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 3).toLowerCase() === peg$c80) {
              s0 = input.substr(peg$currPos, 3);
              peg$currPos += 3;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c81); }
            }
            if (s0 === peg$FAILED) {
              if (input.substr(peg$currPos, 3).toLowerCase() === peg$c82) {
                s0 = input.substr(peg$currPos, 3);
                peg$currPos += 3;
              } else {
                s0 = peg$FAILED;
                if (peg$silentFails === 0) { peg$fail(peg$c83); }
              }
              if (s0 === peg$FAILED) {
                if (input.substr(peg$currPos, 2).toLowerCase() === peg$c84) {
                  s0 = input.substr(peg$currPos, 2);
                  peg$currPos += 2;
                } else {
                  s0 = peg$FAILED;
                  if (peg$silentFails === 0) { peg$fail(peg$c85); }
                }
              }
            }
          }
        }
      }

      return s0;
    }

    function peg$parsePfund() {
      var s0;

      if (input.substr(peg$currPos, 6).toLowerCase() === peg$c86) {
        s0 = input.substr(peg$currPos, 6);
        peg$currPos += 6;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c87); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 5).toLowerCase() === peg$c88) {
          s0 = input.substr(peg$currPos, 5);
          peg$currPos += 5;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c89); }
        }
      }

      return s0;
    }

    function peg$parsequart() {
      var s0;

      if (input.substr(peg$currPos, 6).toLowerCase() === peg$c90) {
        s0 = input.substr(peg$currPos, 6);
        peg$currPos += 6;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c91); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 5).toLowerCase() === peg$c92) {
          s0 = input.substr(peg$currPos, 5);
          peg$currPos += 5;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c93); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 4).toLowerCase() === peg$c94) {
            s0 = input.substr(peg$currPos, 4);
            peg$currPos += 4;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c95); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 3).toLowerCase() === peg$c96) {
              s0 = input.substr(peg$currPos, 3);
              peg$currPos += 3;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c97); }
            }
            if (s0 === peg$FAILED) {
              if (input.substr(peg$currPos, 3).toLowerCase() === peg$c98) {
                s0 = input.substr(peg$currPos, 3);
                peg$currPos += 3;
              } else {
                s0 = peg$FAILED;
                if (peg$silentFails === 0) { peg$fail(peg$c99); }
              }
              if (s0 === peg$FAILED) {
                if (input.substr(peg$currPos, 2).toLowerCase() === peg$c100) {
                  s0 = input.substr(peg$currPos, 2);
                  peg$currPos += 2;
                } else {
                  s0 = peg$FAILED;
                  if (peg$silentFails === 0) { peg$fail(peg$c101); }
                }
              }
            }
          }
        }
      }

      return s0;
    }

    function peg$parsetablespoon() {
      var s0;

      if (input.substr(peg$currPos, 11).toLowerCase() === peg$c102) {
        s0 = input.substr(peg$currPos, 11);
        peg$currPos += 11;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c103); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 10).toLowerCase() === peg$c104) {
          s0 = input.substr(peg$currPos, 10);
          peg$currPos += 10;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c105); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 5).toLowerCase() === peg$c106) {
            s0 = input.substr(peg$currPos, 5);
            peg$currPos += 5;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c107); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 4).toLowerCase() === peg$c108) {
              s0 = input.substr(peg$currPos, 4);
              peg$currPos += 4;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c109); }
            }
            if (s0 === peg$FAILED) {
              if (input.substr(peg$currPos, 4).toLowerCase() === peg$c110) {
                s0 = input.substr(peg$currPos, 4);
                peg$currPos += 4;
              } else {
                s0 = peg$FAILED;
                if (peg$silentFails === 0) { peg$fail(peg$c111); }
              }
              if (s0 === peg$FAILED) {
                if (input.substr(peg$currPos, 3).toLowerCase() === peg$c112) {
                  s0 = input.substr(peg$currPos, 3);
                  peg$currPos += 3;
                } else {
                  s0 = peg$FAILED;
                  if (peg$silentFails === 0) { peg$fail(peg$c113); }
                }
                if (s0 === peg$FAILED) {
                  if (input.substr(peg$currPos, 2) === peg$c114) {
                    s0 = peg$c114;
                    peg$currPos += 2;
                  } else {
                    s0 = peg$FAILED;
                    if (peg$silentFails === 0) { peg$fail(peg$c115); }
                  }
                  if (s0 === peg$FAILED) {
                    if (input.charCodeAt(peg$currPos) === 84) {
                      s0 = peg$c116;
                      peg$currPos++;
                    } else {
                      s0 = peg$FAILED;
                      if (peg$silentFails === 0) { peg$fail(peg$c117); }
                    }
                  }
                }
              }
            }
          }
        }
      }

      return s0;
    }

    function peg$parseEsslöffel() {
      var s0;

      if (input.substr(peg$currPos, 19).toLowerCase() === peg$c118) {
        s0 = input.substr(peg$currPos, 19);
        peg$currPos += 19;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c119); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 18).toLowerCase() === peg$c120) {
          s0 = input.substr(peg$currPos, 18);
          peg$currPos += 18;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c121); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 18).toLowerCase() === peg$c122) {
            s0 = input.substr(peg$currPos, 18);
            peg$currPos += 18;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c123); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 17).toLowerCase() === peg$c124) {
              s0 = input.substr(peg$currPos, 17);
              peg$currPos += 17;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c125); }
            }
            if (s0 === peg$FAILED) {
              if (input.substr(peg$currPos, 12).toLowerCase() === peg$c126) {
                s0 = input.substr(peg$currPos, 12);
                peg$currPos += 12;
              } else {
                s0 = peg$FAILED;
                if (peg$silentFails === 0) { peg$fail(peg$c127); }
              }
              if (s0 === peg$FAILED) {
                if (input.substr(peg$currPos, 11).toLowerCase() === peg$c128) {
                  s0 = input.substr(peg$currPos, 11);
                  peg$currPos += 11;
                } else {
                  s0 = peg$FAILED;
                  if (peg$silentFails === 0) { peg$fail(peg$c129); }
                }
                if (s0 === peg$FAILED) {
                  if (input.substr(peg$currPos, 22).toLowerCase() === peg$c130) {
                    s0 = input.substr(peg$currPos, 22);
                    peg$currPos += 22;
                  } else {
                    s0 = peg$FAILED;
                    if (peg$silentFails === 0) { peg$fail(peg$c131); }
                  }
                  if (s0 === peg$FAILED) {
                    if (input.substr(peg$currPos, 21).toLowerCase() === peg$c132) {
                      s0 = input.substr(peg$currPos, 21);
                      peg$currPos += 21;
                    } else {
                      s0 = peg$FAILED;
                      if (peg$silentFails === 0) { peg$fail(peg$c133); }
                    }
                    if (s0 === peg$FAILED) {
                      if (input.substr(peg$currPos, 21).toLowerCase() === peg$c134) {
                        s0 = input.substr(peg$currPos, 21);
                        peg$currPos += 21;
                      } else {
                        s0 = peg$FAILED;
                        if (peg$silentFails === 0) { peg$fail(peg$c135); }
                      }
                      if (s0 === peg$FAILED) {
                        if (input.substr(peg$currPos, 20).toLowerCase() === peg$c136) {
                          s0 = input.substr(peg$currPos, 20);
                          peg$currPos += 20;
                        } else {
                          s0 = peg$FAILED;
                          if (peg$silentFails === 0) { peg$fail(peg$c137); }
                        }
                        if (s0 === peg$FAILED) {
                          if (input.substr(peg$currPos, 16).toLowerCase() === peg$c138) {
                            s0 = input.substr(peg$currPos, 16);
                            peg$currPos += 16;
                          } else {
                            s0 = peg$FAILED;
                            if (peg$silentFails === 0) { peg$fail(peg$c139); }
                          }
                          if (s0 === peg$FAILED) {
                            if (input.substr(peg$currPos, 15).toLowerCase() === peg$c140) {
                              s0 = input.substr(peg$currPos, 15);
                              peg$currPos += 15;
                            } else {
                              s0 = peg$FAILED;
                              if (peg$silentFails === 0) { peg$fail(peg$c141); }
                            }
                            if (s0 === peg$FAILED) {
                              if (input.substr(peg$currPos, 9).toLowerCase() === peg$c142) {
                                s0 = input.substr(peg$currPos, 9);
                                peg$currPos += 9;
                              } else {
                                s0 = peg$FAILED;
                                if (peg$silentFails === 0) { peg$fail(peg$c143); }
                              }
                              if (s0 === peg$FAILED) {
                                if (input.substr(peg$currPos, 8).toLowerCase() === peg$c144) {
                                  s0 = input.substr(peg$currPos, 8);
                                  peg$currPos += 8;
                                } else {
                                  s0 = peg$FAILED;
                                  if (peg$silentFails === 0) { peg$fail(peg$c145); }
                                }
                                if (s0 === peg$FAILED) {
                                  if (input.substr(peg$currPos, 9).toLowerCase() === peg$c146) {
                                    s0 = input.substr(peg$currPos, 9);
                                    peg$currPos += 9;
                                  } else {
                                    s0 = peg$FAILED;
                                    if (peg$silentFails === 0) { peg$fail(peg$c147); }
                                  }
                                  if (s0 === peg$FAILED) {
                                    if (input.substr(peg$currPos, 8).toLowerCase() === peg$c148) {
                                      s0 = input.substr(peg$currPos, 8);
                                      peg$currPos += 8;
                                    } else {
                                      s0 = peg$FAILED;
                                      if (peg$silentFails === 0) { peg$fail(peg$c149); }
                                    }
                                    if (s0 === peg$FAILED) {
                                      if (input.substr(peg$currPos, 7).toLowerCase() === peg$c150) {
                                        s0 = input.substr(peg$currPos, 7);
                                        peg$currPos += 7;
                                      } else {
                                        s0 = peg$FAILED;
                                        if (peg$silentFails === 0) { peg$fail(peg$c151); }
                                      }
                                      if (s0 === peg$FAILED) {
                                        if (input.substr(peg$currPos, 6).toLowerCase() === peg$c152) {
                                          s0 = input.substr(peg$currPos, 6);
                                          peg$currPos += 6;
                                        } else {
                                          s0 = peg$FAILED;
                                          if (peg$silentFails === 0) { peg$fail(peg$c153); }
                                        }
                                        if (s0 === peg$FAILED) {
                                          if (input.substr(peg$currPos, 2).toLowerCase() === peg$c154) {
                                            s0 = input.substr(peg$currPos, 2);
                                            peg$currPos += 2;
                                          } else {
                                            s0 = peg$FAILED;
                                            if (peg$silentFails === 0) { peg$fail(peg$c155); }
                                          }
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }

      return s0;
    }

    function peg$parseteaspoon() {
      var s0;

      if (input.substr(peg$currPos, 9).toLowerCase() === peg$c156) {
        s0 = input.substr(peg$currPos, 9);
        peg$currPos += 9;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c157); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 8).toLowerCase() === peg$c158) {
          s0 = input.substr(peg$currPos, 8);
          peg$currPos += 8;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c159); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 4).toLowerCase() === peg$c160) {
            s0 = input.substr(peg$currPos, 4);
            peg$currPos += 4;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c161); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 3).toLowerCase() === peg$c162) {
              s0 = input.substr(peg$currPos, 3);
              peg$currPos += 3;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c163); }
            }
            if (s0 === peg$FAILED) {
              if (input.substr(peg$currPos, 2) === peg$c42) {
                s0 = peg$c42;
                peg$currPos += 2;
              } else {
                s0 = peg$FAILED;
                if (peg$silentFails === 0) { peg$fail(peg$c43); }
              }
              if (s0 === peg$FAILED) {
                if (input.charCodeAt(peg$currPos) === 116) {
                  s0 = peg$c164;
                  peg$currPos++;
                } else {
                  s0 = peg$FAILED;
                  if (peg$silentFails === 0) { peg$fail(peg$c165); }
                }
              }
            }
          }
        }
      }

      return s0;
    }

    function peg$parseTeelöffel() {
      var s0;

      if (input.substr(peg$currPos, 19).toLowerCase() === peg$c166) {
        s0 = input.substr(peg$currPos, 19);
        peg$currPos += 19;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c167); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 18).toLowerCase() === peg$c168) {
          s0 = input.substr(peg$currPos, 18);
          peg$currPos += 18;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c169); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 11).toLowerCase() === peg$c170) {
            s0 = input.substr(peg$currPos, 11);
            peg$currPos += 11;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c171); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 12).toLowerCase() === peg$c172) {
              s0 = input.substr(peg$currPos, 12);
              peg$currPos += 12;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c173); }
            }
            if (s0 === peg$FAILED) {
              if (input.substr(peg$currPos, 22).toLowerCase() === peg$c174) {
                s0 = input.substr(peg$currPos, 22);
                peg$currPos += 22;
              } else {
                s0 = peg$FAILED;
                if (peg$silentFails === 0) { peg$fail(peg$c175); }
              }
              if (s0 === peg$FAILED) {
                if (input.substr(peg$currPos, 15).toLowerCase() === peg$c176) {
                  s0 = input.substr(peg$currPos, 15);
                  peg$currPos += 15;
                } else {
                  s0 = peg$FAILED;
                  if (peg$silentFails === 0) { peg$fail(peg$c177); }
                }
                if (s0 === peg$FAILED) {
                  if (input.substr(peg$currPos, 21).toLowerCase() === peg$c178) {
                    s0 = input.substr(peg$currPos, 21);
                    peg$currPos += 21;
                  } else {
                    s0 = peg$FAILED;
                    if (peg$silentFails === 0) { peg$fail(peg$c179); }
                  }
                  if (s0 === peg$FAILED) {
                    if (input.substr(peg$currPos, 14).toLowerCase() === peg$c180) {
                      s0 = input.substr(peg$currPos, 14);
                      peg$currPos += 14;
                    } else {
                      s0 = peg$FAILED;
                      if (peg$silentFails === 0) { peg$fail(peg$c181); }
                    }
                    if (s0 === peg$FAILED) {
                      if (input.substr(peg$currPos, 16).toLowerCase() === peg$c182) {
                        s0 = input.substr(peg$currPos, 16);
                        peg$currPos += 16;
                      } else {
                        s0 = peg$FAILED;
                        if (peg$silentFails === 0) { peg$fail(peg$c183); }
                      }
                      if (s0 === peg$FAILED) {
                        if (input.substr(peg$currPos, 9).toLowerCase() === peg$c184) {
                          s0 = input.substr(peg$currPos, 9);
                          peg$currPos += 9;
                        } else {
                          s0 = peg$FAILED;
                          if (peg$silentFails === 0) { peg$fail(peg$c185); }
                        }
                        if (s0 === peg$FAILED) {
                          if (input.substr(peg$currPos, 9).toLowerCase() === peg$c186) {
                            s0 = input.substr(peg$currPos, 9);
                            peg$currPos += 9;
                          } else {
                            s0 = peg$FAILED;
                            if (peg$silentFails === 0) { peg$fail(peg$c187); }
                          }
                          if (s0 === peg$FAILED) {
                            if (input.substr(peg$currPos, 7).toLowerCase() === peg$c188) {
                              s0 = input.substr(peg$currPos, 7);
                              peg$currPos += 7;
                            } else {
                              s0 = peg$FAILED;
                              if (peg$silentFails === 0) { peg$fail(peg$c189); }
                            }
                            if (s0 === peg$FAILED) {
                              if (input.substr(peg$currPos, 6).toLowerCase() === peg$c190) {
                                s0 = input.substr(peg$currPos, 6);
                                peg$currPos += 6;
                              } else {
                                s0 = peg$FAILED;
                                if (peg$silentFails === 0) { peg$fail(peg$c191); }
                              }
                              if (s0 === peg$FAILED) {
                                if (input.substr(peg$currPos, 2).toLowerCase() === peg$c192) {
                                  s0 = input.substr(peg$currPos, 2);
                                  peg$currPos += 2;
                                } else {
                                  s0 = peg$FAILED;
                                  if (peg$silentFails === 0) { peg$fail(peg$c193); }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }

      return s0;
    }

    function peg$parseMesserspitze() {
      var s0;

      if (input.substr(peg$currPos, 13).toLowerCase() === peg$c194) {
        s0 = input.substr(peg$currPos, 13);
        peg$currPos += 13;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c195); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 12).toLowerCase() === peg$c196) {
          s0 = input.substr(peg$currPos, 12);
          peg$currPos += 12;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c197); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 4).toLowerCase() === peg$c198) {
            s0 = input.substr(peg$currPos, 4);
            peg$currPos += 4;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c199); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 3).toLowerCase() === peg$c200) {
              s0 = input.substr(peg$currPos, 3);
              peg$currPos += 3;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c201); }
            }
          }
        }
      }

      return s0;
    }

    function peg$parsePrise() {
      var s0;

      if (input.substr(peg$currPos, 6).toLowerCase() === peg$c202) {
        s0 = input.substr(peg$currPos, 6);
        peg$currPos += 6;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c203); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 5).toLowerCase() === peg$c204) {
          s0 = input.substr(peg$currPos, 5);
          peg$currPos += 5;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c205); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 3).toLowerCase() === peg$c206) {
            s0 = input.substr(peg$currPos, 3);
            peg$currPos += 3;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c207); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 2).toLowerCase() === peg$c208) {
              s0 = input.substr(peg$currPos, 2);
              peg$currPos += 2;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c209); }
            }
          }
        }
      }

      return s0;
    }

    function peg$parseSchuss() {
      var s0;

      if (input.substr(peg$currPos, 7).toLowerCase() === peg$c210) {
        s0 = input.substr(peg$currPos, 7);
        peg$currPos += 7;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c211); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 6).toLowerCase() === peg$c212) {
          s0 = input.substr(peg$currPos, 6);
          peg$currPos += 6;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c213); }
        }
      }

      return s0;
    }

    function peg$parseSpritzer() {
      var s0;

      if (input.substr(peg$currPos, 8).toLowerCase() === peg$c214) {
        s0 = input.substr(peg$currPos, 8);
        peg$currPos += 8;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c215); }
      }

      return s0;
    }

    function peg$parseTropfen() {
      var s0;

      if (input.substr(peg$currPos, 7).toLowerCase() === peg$c216) {
        s0 = input.substr(peg$currPos, 7);
        peg$currPos += 7;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c217); }
      }

      return s0;
    }

    function peg$parseBund() {
      var s0;

      if (input.substr(peg$currPos, 5).toLowerCase() === peg$c218) {
        s0 = input.substr(peg$currPos, 5);
        peg$currPos += 5;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c219); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 5).toLowerCase() === peg$c220) {
          s0 = input.substr(peg$currPos, 5);
          peg$currPos += 5;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c221); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 4).toLowerCase() === peg$c222) {
            s0 = input.substr(peg$currPos, 4);
            peg$currPos += 4;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c223); }
          }
        }
      }

      return s0;
    }

    function peg$parsemetric_unit() {
      var s0;

      s0 = peg$parsegram();
      if (s0 === peg$FAILED) {
        s0 = peg$parsekilogram();
        if (s0 === peg$FAILED) {
          s0 = peg$parseliter();
          if (s0 === peg$FAILED) {
            s0 = peg$parsemilligram();
            if (s0 === peg$FAILED) {
              s0 = peg$parsemilliliter();
            }
          }
        }
      }

      return s0;
    }

    function peg$parsegram() {
      var s0;

      if (input.substr(peg$currPos, 5).toLowerCase() === peg$c224) {
        s0 = input.substr(peg$currPos, 5);
        peg$currPos += 5;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c225); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 4).toLowerCase() === peg$c226) {
          s0 = input.substr(peg$currPos, 4);
          peg$currPos += 4;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c227); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 3).toLowerCase() === peg$c228) {
            s0 = input.substr(peg$currPos, 3);
            peg$currPos += 3;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c229); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 2).toLowerCase() === peg$c230) {
              s0 = input.substr(peg$currPos, 2);
              peg$currPos += 2;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c231); }
            }
            if (s0 === peg$FAILED) {
              if (input.substr(peg$currPos, 2).toLowerCase() === peg$c232) {
                s0 = input.substr(peg$currPos, 2);
                peg$currPos += 2;
              } else {
                s0 = peg$FAILED;
                if (peg$silentFails === 0) { peg$fail(peg$c233); }
              }
              if (s0 === peg$FAILED) {
                if (input.substr(peg$currPos, 1).toLowerCase() === peg$c234) {
                  s0 = input.charAt(peg$currPos);
                  peg$currPos++;
                } else {
                  s0 = peg$FAILED;
                  if (peg$silentFails === 0) { peg$fail(peg$c235); }
                }
              }
            }
          }
        }
      }

      return s0;
    }

    function peg$parsekilogram() {
      var s0;

      if (input.substr(peg$currPos, 9).toLowerCase() === peg$c236) {
        s0 = input.substr(peg$currPos, 9);
        peg$currPos += 9;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c237); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 8).toLowerCase() === peg$c238) {
          s0 = input.substr(peg$currPos, 8);
          peg$currPos += 8;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c239); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 3).toLowerCase() === peg$c240) {
            s0 = input.substr(peg$currPos, 3);
            peg$currPos += 3;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c241); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 2).toLowerCase() === peg$c242) {
              s0 = input.substr(peg$currPos, 2);
              peg$currPos += 2;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c243); }
            }
          }
        }
      }

      return s0;
    }

    function peg$parseliter() {
      var s0;

      if (input.substr(peg$currPos, 6).toLowerCase() === peg$c244) {
        s0 = input.substr(peg$currPos, 6);
        peg$currPos += 6;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c245); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 5).toLowerCase() === peg$c246) {
          s0 = input.substr(peg$currPos, 5);
          peg$currPos += 5;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c247); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 2).toLowerCase() === peg$c248) {
            s0 = input.substr(peg$currPos, 2);
            peg$currPos += 2;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c249); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 1).toLowerCase() === peg$c250) {
              s0 = input.charAt(peg$currPos);
              peg$currPos++;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c251); }
            }
          }
        }
      }

      return s0;
    }

    function peg$parsemilligram() {
      var s0;

      if (input.substr(peg$currPos, 10).toLowerCase() === peg$c252) {
        s0 = input.substr(peg$currPos, 10);
        peg$currPos += 10;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c253); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 9).toLowerCase() === peg$c254) {
          s0 = input.substr(peg$currPos, 9);
          peg$currPos += 9;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c255); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 3).toLowerCase() === peg$c256) {
            s0 = input.substr(peg$currPos, 3);
            peg$currPos += 3;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c257); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 2).toLowerCase() === peg$c258) {
              s0 = input.substr(peg$currPos, 2);
              peg$currPos += 2;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c259); }
            }
          }
        }
      }

      return s0;
    }

    function peg$parsemilliliter() {
      var s0;

      if (input.substr(peg$currPos, 11).toLowerCase() === peg$c260) {
        s0 = input.substr(peg$currPos, 11);
        peg$currPos += 11;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c261); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 10).toLowerCase() === peg$c262) {
          s0 = input.substr(peg$currPos, 10);
          peg$currPos += 10;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c263); }
        }
        if (s0 === peg$FAILED) {
          if (input.substr(peg$currPos, 3).toLowerCase() === peg$c264) {
            s0 = input.substr(peg$currPos, 3);
            peg$currPos += 3;
          } else {
            s0 = peg$FAILED;
            if (peg$silentFails === 0) { peg$fail(peg$c265); }
          }
          if (s0 === peg$FAILED) {
            if (input.substr(peg$currPos, 2).toLowerCase() === peg$c266) {
              s0 = input.substr(peg$currPos, 2);
              peg$currPos += 2;
            } else {
              s0 = peg$FAILED;
              if (peg$silentFails === 0) { peg$fail(peg$c267); }
            }
          }
        }
      }

      return s0;
    }

    function peg$parseimprecise_unit() {
      var s0;

      s0 = peg$parsedash();
      if (s0 === peg$FAILED) {
        s0 = peg$parsehandful();
        if (s0 === peg$FAILED) {
          s0 = peg$parsepinch();
          if (s0 === peg$FAILED) {
            s0 = peg$parsetouch();
            if (s0 === peg$FAILED) {
              s0 = peg$parseetwas();
            }
          }
        }
      }

      return s0;
    }

    function peg$parsedash() {
      var s0;

      if (input.substr(peg$currPos, 6).toLowerCase() === peg$c268) {
        s0 = input.substr(peg$currPos, 6);
        peg$currPos += 6;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c269); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 4).toLowerCase() === peg$c270) {
          s0 = input.substr(peg$currPos, 4);
          peg$currPos += 4;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c271); }
        }
      }

      return s0;
    }

    function peg$parsehandful() {
      var s0;

      if (input.substr(peg$currPos, 8).toLowerCase() === peg$c272) {
        s0 = input.substr(peg$currPos, 8);
        peg$currPos += 8;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c273); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 7).toLowerCase() === peg$c274) {
          s0 = input.substr(peg$currPos, 7);
          peg$currPos += 7;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c275); }
        }
      }

      return s0;
    }

    function peg$parsepinch() {
      var s0;

      if (input.substr(peg$currPos, 7).toLowerCase() === peg$c276) {
        s0 = input.substr(peg$currPos, 7);
        peg$currPos += 7;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c277); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 5).toLowerCase() === peg$c278) {
          s0 = input.substr(peg$currPos, 5);
          peg$currPos += 5;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c279); }
        }
      }

      return s0;
    }

    function peg$parsetouch() {
      var s0;

      if (input.substr(peg$currPos, 7).toLowerCase() === peg$c280) {
        s0 = input.substr(peg$currPos, 7);
        peg$currPos += 7;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c281); }
      }
      if (s0 === peg$FAILED) {
        if (input.substr(peg$currPos, 5).toLowerCase() === peg$c282) {
          s0 = input.substr(peg$currPos, 5);
          peg$currPos += 5;
        } else {
          s0 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c283); }
        }
      }

      return s0;
    }

    function peg$parseetwas() {
      var s0;

      if (input.substr(peg$currPos, 5).toLowerCase() === peg$c284) {
        s0 = input.substr(peg$currPos, 5);
        peg$currPos += 5;
      } else {
        s0 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c285); }
      }

      return s0;
    }

    peg$result = peg$startRuleFunction();

    if (peg$result !== peg$FAILED && peg$currPos === input.length) {
      return peg$result;
    } else {
      if (peg$result !== peg$FAILED && peg$currPos < input.length) {
        peg$fail({ type: "end", description: "end of input" });
      }

      throw peg$buildException(
        null,
        peg$maxFailExpected,
        peg$maxFailPos < input.length ? input.charAt(peg$maxFailPos) : null,
        peg$maxFailPos < input.length
          ? peg$computeLocation(peg$maxFailPos, peg$maxFailPos + 1)
          : peg$computeLocation(peg$maxFailPos, peg$maxFailPos)
      );
    }
  }

  return {
    SyntaxError: peg$SyntaxError,
    parse:       peg$parse
  };
})();
